src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+            'CB_UB_GE_n' : sort hierarchically by CB, UB, GE, and query_name
+            'n' : only sort by query_name
+
+        Returns
+        -------
+        None
+        """
+        assert 2 <= max_gene_hits_per_multiple_gene_alignments < self.max_genes
+        assert num_duplicates >= 0
+        assert num_missing_some_tags >= 0
+        assert num_multiple_gene_alignments >= 0
+
+        # generate synthetic count matrix and corresponding simulated records
+        synthetic_data_bundle = self.generate_synthetic_counts_and_alignment_tags(
+            num_duplicates,
+            num_missing_some_tags,
+            num_multiple_gene_alignments,
+            max_gene_hits_per_multiple_gene_alignments)
+        records = [record for record in synthetic_data_bundle.get_bam_records_generator()]
+
+        if query_sort_order == 'random':
+            # shuffle records
+            self.rng.shuffle(records)
+
+        elif query_sort_order == 'CB_UB_GE_n':
+            # sort by (CB, GE, UB, query_name)
+            def record_sort_key(record: pysam.AlignedSegment):
+                query_name = record.query_name
+                try:
+                    cell_barcode = record.get_tag(self.cell_barcode_tag)
+                except KeyError:
+                    cell_barcode = 'N'
+                try:
+                    molecule_barcode = record.get_tag(self.molecule_barcode_tag)
+                except KeyError:
+                    molecule_barcode = 'N'
+                try:
+                    gene_name = record.get_tag(self.gene_name_tag)
+                except KeyError:
+                    gene_name = 'N'
+                return cell_barcode, molecule_barcode, gene_name, query_name
+            records = sorted(records, key=record_sort_key)
+
+        elif query_sort_order == 'n':
+            def record_sort_key(record: pysam.AlignedSegment):
+                return record.query_name
+            records = sorted(records, key=record_sort_key)
+
+        else:
+            raise Exception(f'Unknown query sort order {query_sort_order}')
+
+        # write BAM file
+        with pysam.AlignmentFile(os.path.join(output_path, self.bam_output_filename),
+                                 mode='wb',
+                                 reference_names=[SyntheticDataBundle.synthetic_sequence_name],
+                                 reference_lengths=[SyntheticDataBundle.synthetic_sequence_length]) as bo:
+            for record in records:
+                bo.write(record)
+
+        # write count matrix, row index, and col index
+        np.save(os.path.join(output_path, self.count_matrix_output_filename), synthetic_data_bundle.count_matrix)
+        np.save(os.path.join(output_path, self.row_index_output_filename), synthetic_data_bundle.row_index)
+        np.save(os.path.join(output_path, self.col_index_output_filename), synthetic_data_bundle.col_index)
+
+    def _generate_random_count_matrix(self) -> np.ndarray:
```

Comment:
The code here is clear but I think a doc block here and for the below functions (for developers) would be nice. What do you think?

Reply:
Done.